File: javalin_questions/APQuiz09Test.java
/**
INSTRUCTIONS: 
    The APQuiz09Test class provides JUnit tests for the methods 
    in the APQuiz09 class.



COMPILE & EXECUTE & CLEANUP (Java):

     javac     -d out           -cp junit-1.9.3.jar     APQuiz09*.java
     java -jar junit-1.9.3.jar  --class-path "out"      --scan-classpath
     rm -rf out/

COMPILE & EXECUTE & CLEANUP (Java x Kotlin):

     javac     -d out   -cp junit-1.9.3.jar             APQuiz09*.java
     kotlinc    -d out                                  APQuiz09*.kt
     java -jar junit-1.9.3.jar --class-path "./out:$KOTLIN_HOME/lib/kotlin-stdlib.jar" --scan-classpath
     rm -rf out/


TIP:
     - On Windows, use ';' as the classpath separator
     - On Mac/Linux, use ':' as the classpath separator
     - java   `-cp` (Mac/Linux) flag must sepcify absolute paths.
     - kotlin `-cp` (Windows) flag accepts relative paths.
 */

// File: APQuiz09Test.java
// JDK-17: https://developers.redhat.com/products/openjdk/download?sc_cid=701f2000000RWTnAAO
// JUnit: https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter/5.10.0
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import java.util.Arrays;
import java.util.List;

public class APQuiz09Test {
    
    /**
     * < Q1 >
     * Tests the shiftLeft method.
     */
    @Test
    public void testShiftLeft() {
        assertArrayEquals(new int[]{2, 5, 3, 6}, APQuiz09.shiftLeft(new int[]{6, 2, 5, 3}));
        assertArrayEquals(new int[]{2, 1}, APQuiz09.shiftLeft(new int[]{1, 2}));
        assertArrayEquals(new int[]{1}, APQuiz09.shiftLeft(new int[]{1}));
        assertArrayEquals(new int[]{}, APQuiz09.shiftLeft(new int[]{}));
        assertArrayEquals(new int[]{1, 2, 2, 4, 1}, APQuiz09.shiftLeft(new int[]{1, 1, 2, 2, 4}));
        assertArrayEquals(new int[]{1, 1, 1}, APQuiz09.shiftLeft(new int[]{1, 1, 1}));
        assertArrayEquals(new int[]{2, 3, 1}, APQuiz09.shiftLeft(new int[]{1, 2, 3}));
    }

    /**
     * < Q2 >
     * Tests the tenRun method.
     */
    @Test
    public void testTenRun() {
        assertArrayEquals(new int[]{2, 10, 10, 10, 20, 20}, APQuiz09.tenRun(new int[]{2, 10, 3, 4, 20, 5}));
        assertArrayEquals(new int[]{10, 10, 20, 20}, APQuiz09.tenRun(new int[]{10, 1, 20, 2}));
        assertArrayEquals(new int[]{10, 10, 10, 20}, APQuiz09.tenRun(new int[]{10, 1, 9, 20}));
        assertArrayEquals(new int[]{1, 2, 50, 50}, APQuiz09.tenRun(new int[]{1, 2, 50, 1}));
        assertArrayEquals(new int[]{1, 20, 50, 50}, APQuiz09.tenRun(new int[]{1, 20, 50, 1}));
        assertArrayEquals(new int[]{10, 10}, APQuiz09.tenRun(new int[]{10, 10}));
        assertArrayEquals(new int[]{10, 10}, APQuiz09.tenRun(new int[]{10, 2}));
        assertArrayEquals(new int[]{0, 0}, APQuiz09.tenRun(new int[]{0, 2}));
        assertArrayEquals(new int[]{1, 2}, APQuiz09.tenRun(new int[]{1, 2}));
        assertArrayEquals(new int[]{1}, APQuiz09.tenRun(new int[]{1}));
        assertArrayEquals(new int[]{}, APQuiz09.tenRun(new int[]{}));
    }

    /**
     * < Q3 >
     * Tests the pre4 method.
     */
    @Test
    public void testPre4() {
        assertArrayEquals(new int[]{1, 2}, APQuiz09.pre4(new int[]{1, 2, 4, 1}));
        assertArrayEquals(new int[]{3, 1}, APQuiz09.pre4(new int[]{3, 1, 4}));
        assertArrayEquals(new int[]{1}, APQuiz09.pre4(new int[]{1, 4, 4}));
        assertArrayEquals(new int[]{1}, APQuiz09.pre4(new int[]{1, 4, 4, 2}));
        assertArrayEquals(new int[]{1, 3}, APQuiz09.pre4(new int[]{1, 3, 4, 2, 4}));
        assertArrayEquals(new int[]{}, APQuiz09.pre4(new int[]{4, 4}));
        assertArrayEquals(new int[]{3, 3}, APQuiz09.pre4(new int[]{3, 3, 4}));
        assertArrayEquals(new int[]{1, 2, 1}, APQuiz09.pre4(new int[]{1, 2, 1, 4}));
        assertArrayEquals(new int[]{2, 1}, APQuiz09.pre4(new int[]{2, 1, 4, 2}));
        assertArrayEquals(new int[]{2, 1, 2, 1}, APQuiz09.pre4(new int[]{2, 1, 2, 1, 4, 2}));
    }

    /**
     * < Q4 >
     * Tests the blackjack method.
     */
    @Test
    public void testBlackjack() {
        assertEquals(21, APQuiz09.blackjack(19, 21));
        assertEquals(21, APQuiz09.blackjack(21, 19));
        assertEquals(19, APQuiz09.blackjack(19, 22));
        assertEquals(19, APQuiz09.blackjack(22, 19));
        assertEquals(0, APQuiz09.blackjack(22, 50));
        assertEquals(0, APQuiz09.blackjack(22, 22));
        assertEquals(1, APQuiz09.blackjack(33, 1));
        assertEquals(2, APQuiz09.blackjack(1, 2));
        assertEquals(0, APQuiz09.blackjack(34, 33));
        assertEquals(19, APQuiz09.blackjack(17, 19));
        assertEquals(18, APQuiz09.blackjack(18, 17));
        assertEquals(16, APQuiz09.blackjack(16, 23));
        assertEquals(4, APQuiz09.blackjack(3, 4));
        assertEquals(3, APQuiz09.blackjack(3, 2));
        assertEquals(21, APQuiz09.blackjack(21, 20));
    }

    /**
     * < Q5 >
     * Tests the maxBlock method.
     */
    @Test
    public void testMaxBlock() {
        assertEquals(2, APQuiz09.maxBlock("hoopla"));
        assertEquals(3, APQuiz09.maxBlock("abbCCCddBBBxx"));
        assertEquals(0, APQuiz09.maxBlock(""));
        assertEquals(1, APQuiz09.maxBlock("xyz"));
        assertEquals(2, APQuiz09.maxBlock("xxyz"));
        assertEquals(2, APQuiz09.maxBlock("xyzz"));
        assertEquals(3, APQuiz09.maxBlock("abbbcbbbxbbbx"));
        assertEquals(3, APQuiz09.maxBlock("XXBBBbbxx"));
        assertEquals(4, APQuiz09.maxBlock("XXBBBBbbxx"));
        assertEquals(4, APQuiz09.maxBlock("XXBBBbbxxXXXX"));
        assertEquals(4, APQuiz09.maxBlock("XX2222BBBbbXX2222"));
    }

    /**
     * < Q6 >
     * Tests the seriesUp method.
     */
    @Test
    public void testSeriesUp() {
        assertArrayEquals(new int[]{1, 1, 2, 1, 2, 3}, APQuiz09.seriesUp(3));
        assertArrayEquals(new int[]{1, 1, 2, 1, 2, 3, 1, 2, 3, 4}, APQuiz09.seriesUp(4));
        assertArrayEquals(new int[]{1, 1, 2}, APQuiz09.seriesUp(2));
        assertArrayEquals(new int[]{1}, APQuiz09.seriesUp(1));
        assertArrayEquals(new int[]{}, APQuiz09.seriesUp(0));
        assertArrayEquals(new int[]{
            1, 1, 2, 1, 2, 3, 1, 2, 3, 4,
            1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 6
        }, APQuiz09.seriesUp(6));
    }

    /**
     * < Q7 >
     * Tests the oneTwo method.
     */
    @Test
    public void testOneTwo() {
        assertEquals("bca", APQuiz09.oneTwo("abc"));
        assertEquals("cat", APQuiz09.oneTwo("tca"));
        assertEquals("catdog", APQuiz09.oneTwo("tcagdo"));
        assertEquals("hocolctea", APQuiz09.oneTwo("chocolate"));
        assertEquals("231564897", APQuiz09.oneTwo("1234567890"));
        assertEquals("abxabxabxabxabxabxabx", APQuiz09.oneTwo("xabxabxabxabxabxabxab"));
        assertEquals("bcaefd", APQuiz09.oneTwo("abcdefx"));
        assertEquals("bcaefd", APQuiz09.oneTwo("abcdefxy"));
        assertEquals("bcaefdyzx", APQuiz09.oneTwo("abcdefxyz"));
        assertEquals("", APQuiz09.oneTwo(""));
        assertEquals("", APQuiz09.oneTwo("x"));
        assertEquals("", APQuiz09.oneTwo("xy"));
        assertEquals("yzx", APQuiz09.oneTwo("xyz"));
        String longInput = "abcdefghijklkmnopqrstuvwxyz1234567890";
        String longExpected = "bcaefdhigkljmnkpqostrvwuyzx231564897";
        assertEquals(longExpected, APQuiz09.oneTwo(longInput));
        assertEquals(longExpected, APQuiz09.oneTwo(longInput.substring(0, longInput.length() - 1)));
        assertEquals(longExpected.substring(0, longExpected.length() - 3), APQuiz09.oneTwo(longInput.substring(0, longInput.length() - 2)));
    }

    /**
     * < Q8 >
     * Tests the zipZap method.
     */
    @Test
    public void testZipZap() {
        assertEquals("zpXzp", APQuiz09.zipZap("zipXzap"));
        assertEquals("zpzp", APQuiz09.zipZap("zopzop"));
        assertEquals("zzzpzp", APQuiz09.zipZap("zzzopzop"));
        assertEquals("zibzp", APQuiz09.zipZap("zibzap"));
        assertEquals("zp", APQuiz09.zipZap("zip"));
        assertEquals("zi", APQuiz09.zipZap("zi"));
        assertEquals("z", APQuiz09.zipZap("z"));
        assertEquals("", APQuiz09.zipZap(""));
        assertEquals("zp", APQuiz09.zipZap("zzp"));
        assertEquals("abcppp", APQuiz09.zipZap("abcppp"));
        assertEquals("azbcppp", APQuiz09.zipZap("azbcppp"));
        assertEquals("azbcpzp", APQuiz09.zipZap("azbcpzpp"));
    }

    /**
     * < Q9 >
     * Tests the post4 method.
     */
    @Test
    public void testPost4() {
        assertArrayEquals(new int[]{1, 2}, APQuiz09.post4(new int[]{2, 4, 1, 2}));
        assertArrayEquals(new int[]{2}, APQuiz09.post4(new int[]{4, 1, 4, 2}));
        assertArrayEquals(new int[]{1, 2, 3}, APQuiz09.post4(new int[]{4, 4, 1, 2, 3}));
        assertArrayEquals(new int[]{2}, APQuiz09.post4(new int[]{4, 2}));
        assertArrayEquals(new int[]{3}, APQuiz09.post4(new int[]{4, 4, 3}));
        assertArrayEquals(new int[]{}, APQuiz09.post4(new int[]{4, 4}));
        assertArrayEquals(new int[]{}, APQuiz09.post4(new int[]{4}));
        assertArrayEquals(new int[]{3, 2}, APQuiz09.post4(new int[]{2, 4, 1, 4, 3, 2}));
        assertArrayEquals(new int[]{2, 2, 2}, APQuiz09.post4(new int[]{4, 1, 4, 2, 2, 2}));
        assertArrayEquals(new int[]{3, 2}, APQuiz09.post4(new int[]{3, 4, 3, 2}));
    }

    /**
     * < Q10 >
     * Tests the notAlone method.
     */
    @Test
    public void testNotAlone() {
        assertArrayEquals(new int[]{1, 3, 3}, APQuiz09.notAlone(new int[]{1, 2, 3}, 2));
        assertArrayEquals(new int[]{1, 3, 3, 5, 5, 2}, APQuiz09.notAlone(new int[]{1, 2, 3, 2, 5, 2}, 2));
        assertArrayEquals(new int[]{3, 4}, APQuiz09.notAlone(new int[]{3, 4}, 3));
        assertArrayEquals(new int[]{3, 3}, APQuiz09.notAlone(new int[]{3, 3}, 3));
        assertArrayEquals(new int[]{1, 3, 3, 2}, APQuiz09.notAlone(new int[]{1, 3, 1, 2}, 1));
        assertArrayEquals(new int[]{3}, APQuiz09.notAlone(new int[]{3}, 3));
        assertArrayEquals(new int[]{}, APQuiz09.notAlone(new int[]{}, 3));
        assertArrayEquals(new int[]{7, 7, 6}, APQuiz09.notAlone(new int[]{7, 1, 6}, 1));
        assertArrayEquals(new int[]{1, 1, 1}, APQuiz09.notAlone(new int[]{1, 1, 1}, 1));
        assertArrayEquals(new int[]{1, 1, 1, 2}, APQuiz09.notAlone(new int[]{1, 1, 1, 2}, 1));
    }

    /**
     * < Q11 >
     * Tests the groupNoAdj method.
     */
    @Test
    public void testGroupNoAdj() {
        assertTrue(APQuiz09.groupNoAdj(0, new int[]{2, 5, 10, 4}, 12));
        assertFalse(APQuiz09.groupNoAdj(0, new int[]{2, 5, 10, 4}, 14));
        assertFalse(APQuiz09.groupNoAdj(0, new int[]{2, 5, 10, 4}, 7));
        assertTrue(APQuiz09.groupNoAdj(0, new int[]{2, 5, 10, 4, 2}, 7));
        assertTrue(APQuiz09.groupNoAdj(0, new int[]{2, 5, 10, 4}, 9));
        assertTrue(APQuiz09.groupNoAdj(0, new int[]{10, 2, 2, 3, 3}, 15));
        assertFalse(APQuiz09.groupNoAdj(0, new int[]{10, 2, 2, 3, 3}, 7));
        assertTrue(APQuiz09.groupNoAdj(0, new int[]{}, 0));
        assertTrue(APQuiz09.groupNoAdj(0, new int[]{1}, 1));
        assertFalse(APQuiz09.groupNoAdj(0, new int[]{9}, 1));
        assertTrue(APQuiz09.groupNoAdj(0, new int[]{9}, 0));
        assertTrue(APQuiz09.groupNoAdj(0, new int[]{5, 10, 4, 1}, 11));
    }

    /**
     * < Q12 >
     * Tests the groupSum5 method.
     */
    @Test
    public void testGroupSum5() {
        assertTrue(APQuiz09.groupSum5(0, new int[]{2, 5, 10, 4}, 19));
        assertTrue(APQuiz09.groupSum5(0, new int[]{2, 5, 10, 4}, 17));
        assertFalse(APQuiz09.groupSum5(0, new int[]{2, 5, 10, 4}, 12));
        assertFalse(APQuiz09.groupSum5(0, new int[]{2, 5, 4, 10}, 12));
        assertFalse(APQuiz09.groupSum5(0, new int[]{3, 5, 1}, 4));
        assertTrue(APQuiz09.groupSum5(0, new int[]{3, 5, 1}, 5));
        assertTrue(APQuiz09.groupSum5(0, new int[]{1, 3, 5}, 5));
        assertFalse(APQuiz09.groupSum5(0, new int[]{3, 5, 1}, 9));
        assertFalse(APQuiz09.groupSum5(0, new int[]{2, 5, 10, 4}, 7));
        assertTrue(APQuiz09.groupSum5(0, new int[]{2, 5, 10, 4}, 15));
        assertFalse(APQuiz09.groupSum5(0, new int[]{2, 5, 10, 4}, 11));
        assertTrue(APQuiz09.groupSum5(0, new int[]{1}, 1));
        assertFalse(APQuiz09.groupSum5(0, new int[]{9}, 1));
        assertTrue(APQuiz09.groupSum5(0, new int[]{9}, 0));
        assertTrue(APQuiz09.groupSum5(0, new int[]{}, 0));
    }

    /**
     * < Q13 >
     * Tests the groupSumClump method.
     */
    @Test
    public void testGroupSumClump() {
        assertTrue(APQuiz09.groupSumClump(0, new int[]{2, 4, 8}, 10));
        assertTrue(APQuiz09.groupSumClump(0, new int[]{1, 2, 4, 8, 1}, 14));
        assertFalse(APQuiz09.groupSumClump(0, new int[]{2, 4, 4, 8}, 14));
        assertTrue(APQuiz09.groupSumClump(0, new int[]{8, 2, 2, 1}, 9));
        assertFalse(APQuiz09.groupSumClump(0, new int[]{8, 2, 2, 1}, 11));
        assertTrue(APQuiz09.groupSumClump(0, new int[]{1}, 1));
        assertFalse(APQuiz09.groupSumClump(0, new int[]{9}, 1));
    }

    /**
     * < Q14 >
     * Tests the splitArray method.
     */
    @Test
    public void testSplitArray() {
        assertTrue(APQuiz09.splitArray(new int[]{2, 2}));
        assertFalse(APQuiz09.splitArray(new int[]{2, 3}));
        assertTrue(APQuiz09.splitArray(new int[]{5, 2, 3}));
        assertFalse(APQuiz09.splitArray(new int[]{5, 2, 2}));
        assertTrue(APQuiz09.splitArray(new int[]{1, 1, 1, 1, 1, 1}));
        assertFalse(APQuiz09.splitArray(new int[]{1, 1, 1, 1, 1}));
        assertTrue(APQuiz09.splitArray(new int[]{}));
        assertFalse(APQuiz09.splitArray(new int[]{1}));
        assertFalse(APQuiz09.splitArray(new int[]{3, 5}));
        assertTrue(APQuiz09.splitArray(new int[]{5, 3, 2}));
        assertTrue(APQuiz09.splitArray(new int[]{2, 2, 10, 10, 1, 1}));
        assertFalse(APQuiz09.splitArray(new int[]{1, 2, 2, 10, 10, 1, 1}));
        assertTrue(APQuiz09.splitArray(new int[]{1, 2, 3, 10, 10, 1, 1}));
    }
}

File: prev/ap-quiz-09.test.js
/**
 * ap-quiz-09.test.js
 *
 * JUnit-style tests for the functions in ap-quiz-09.js
 *
 * INSTRUCTIONS:
 *     This file provides Jest tests for the methods in the ap-quiz-09.js module.
 *
 * DEPENDENCIES:
 *     This test file requires Jest testing framework.
 *     Install dependencies:
 *         npm init -y
 *         npm install --save-dev jest
 *
 * COMPILE & EXECUTE:
 *
 *     Run all tests:
 *         npm test
 *         OR
 *         npx jest ap-quiz-09.test.js
 *
 * TIP:
 *     - Make sure the require path matches your project structure
 *     - If the file is in the same directory, use: require('./ap-quiz-09')
 *     - Jest automatically finds files ending in .test.js or .spec.js
 */
const {
  shiftLeft,
  tenRun,
  pre4,
  blackjack,
  maxBlock,
  seriesUp,
  oneTwo,
  zipZap,
  post4,
  notAlone,
  groupNoAdj,
  groupSum5,
  groupSumClump,
  splitArray
} = require('./ap-quiz-09');

describe('APQuiz09', () => {

  test('shiftLeft', () => {
    expect(shiftLeft([6, 2, 5, 3])).toEqual([2, 5, 3, 6]);
    expect(shiftLeft([1, 2])).toEqual([2, 1]);
    expect(shiftLeft([1])).toEqual([1]);
    expect(shiftLeft([])).toEqual([]);
    expect(shiftLeft([1, 1, 2, 2, 4])).toEqual([1, 2, 2, 4, 1]);
    expect(shiftLeft([1, 1, 1])).toEqual([1, 1, 1]);
    expect(shiftLeft([1, 2, 3])).toEqual([2, 3, 1]);
  });

  test('tenRun', () => {
    expect(tenRun([2, 10, 3, 4, 20, 5])).toEqual([2, 10, 10, 10, 20, 20]);
    expect(tenRun([10, 1, 20, 2])).toEqual([10, 10, 20, 20]);
    expect(tenRun([10, 1, 9, 20])).toEqual([10, 10, 10, 20]);
    expect(tenRun([1, 2, 50, 1])).toEqual([1, 2, 50, 50]);
    expect(tenRun([1, 20, 50, 1])).toEqual([1, 20, 50, 50]);
    expect(tenRun([10, 10])).toEqual([10, 10]);
    expect(tenRun([10, 2])).toEqual([10, 10]);
    expect(tenRun([0, 2])).toEqual([0, 0]);
    expect(tenRun([1, 2])).toEqual([1, 2]);
    expect(tenRun([1])).toEqual([1]);
    expect(tenRun([])).toEqual([]);
  });

  test('pre4', () => {
    expect(pre4([1, 2, 4, 1])).toEqual([1, 2]);
    expect(pre4([3, 1, 4])).toEqual([3, 1]);
    expect(pre4([1, 4, 4])).toEqual([1]);
    expect(pre4([1, 4, 4, 2])).toEqual([1]);
    expect(pre4([1, 3, 4, 2, 4])).toEqual([1, 3]);
    expect(pre4([4, 4])).toEqual([]);
    expect(pre4([3, 3, 4])).toEqual([3, 3]);
    expect(pre4([1, 2, 1, 4])).toEqual([1, 2, 1]);
    expect(pre4([2, 1, 4, 2])).toEqual([2, 1]);
    expect(pre4([2, 1, 2, 1, 4, 2])).toEqual([2, 1, 2, 1]);
  });

  test('blackjack', () => {
    expect(blackjack(19, 21)).toBe(21);
    expect(blackjack(21, 19)).toBe(21);
    expect(blackjack(19, 22)).toBe(19);
    expect(blackjack(22, 19)).toBe(19);
    expect(blackjack(22, 50)).toBe(0);
    expect(blackjack(22, 22)).toBe(0);
    expect(blackjack(33, 1)).toBe(1);
    expect(blackjack(1, 2)).toBe(2);
    expect(blackjack(34, 33)).toBe(0);
    expect(blackjack(17, 19)).toBe(19);
    expect(blackjack(18, 17)).toBe(18);
    expect(blackjack(16, 23)).toBe(16);
    expect(blackjack(3, 4)).toBe(4);
    expect(blackjack(3, 2)).toBe(3);
    expect(blackjack(21, 20)).toBe(21);
  });

  test('maxBlock', () => {
    expect(maxBlock('hoopla')).toBe(2);
    expect(maxBlock('abbCCCddBBBxx')).toBe(3);
    expect(maxBlock('')).toBe(0);
    expect(maxBlock('xyz')).toBe(1);
    expect(maxBlock('xxyz')).toBe(2);
    expect(maxBlock('xyzz')).toBe(2);
    expect(maxBlock('abbbcbbbxbbbx')).toBe(3);
    expect(maxBlock('XXBBBbbxx')).toBe(3);
    expect(maxBlock('XXBBBBbbxx')).toBe(4);
    expect(maxBlock('XXBBBbbxxXXXX')).toBe(4);
    expect(maxBlock('XX2222BBBbbXX2222')).toBe(4);
  });

  test('seriesUp', () => {
    expect(seriesUp(3)).toEqual([1, 1, 2, 1, 2, 3]);
    expect(seriesUp(4)).toEqual([1, 1, 2, 1, 2, 3, 1, 2, 3, 4]);
    expect(seriesUp(2)).toEqual([1, 1, 2]);
    expect(seriesUp(1)).toEqual([1]);
    expect(seriesUp(0)).toEqual([]);
    expect(seriesUp(6)).toEqual([
      1, 1, 2, 1, 2, 3, 1, 2, 3, 4,
      1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 6
    ]);
  });

  test('oneTwo', () => {
    expect(oneTwo('abc')).toBe('bca');
    expect(oneTwo('tca')).toBe('cat');
    expect(oneTwo('tcagdo')).toBe('catdog');
    expect(oneTwo('chocolate')).toBe('hocolctea');
    expect(oneTwo('1234567890')).toBe('231564897');
    expect(oneTwo('xabxabxabxabxabxabxab')).toBe('abxabxabxabxabxabxabx');
    expect(oneTwo('abcdefx')).toBe('bcaefd');
    expect(oneTwo('abcdefxy')).toBe('bcaefd');
    expect(oneTwo('abcdefxyz')).toBe('bcaefdyzx');
    expect(oneTwo('')).toBe('');
    expect(oneTwo('x')).toBe('');
    expect(oneTwo('xy')).toBe('');
    expect(oneTwo('xyz')).toBe('yzx');
    const longInput = 'abcdefghijklkmnopqrstuvwxyz1234567890';
    const longExpected = 'bcaefdhigkljmnkpqostrvwuyzx231564897';
    expect(oneTwo(longInput)).toBe(longExpected);
    expect(oneTwo(longInput.substring(0, longInput.length - 1))).toBe(longExpected);
    expect(oneTwo(longInput.substring(0, longInput.length - 2))).toBe(longExpected.substring(0, longExpected.length - 3));
  });

  test('zipZap', () => {
    expect(zipZap('zipXzap')).toBe('zpXzp');
    expect(zipZap('zopzop')).toBe('zpzp');
    expect(zipZap('zzzopzop')).toBe('zzzpzp');
    expect(zipZap('zibzap')).toBe('zibzp');
    expect(zipZap('zip')).toBe('zp');
    expect(zipZap('zi')).toBe('zi');
    expect(zipZap('z')).toBe('z');
    expect(zipZap('')).toBe('');
    expect(zipZap('zzp')).toBe('zp');
    expect(zipZap('abcppp')).toBe('abcppp');
    expect(zipZap('azbcppp')).toBe('azbcppp');
    expect(zipZap('azbcpzpp')).toBe('azbcpzp');
  });

  test('post4', () => {
    expect(post4([2, 4, 1, 2])).toEqual([1, 2]);
    expect(post4([4, 1, 4, 2])).toEqual([2]);
    expect(post4([4, 4, 1, 2, 3])).toEqual([1, 2, 3]);
    expect(post4([4, 2])).toEqual([2]);
    expect(post4([4, 4, 3])).toEqual([3]);
    expect(post4([4, 4])).toEqual([]);
    expect(post4([4])).toEqual([]);
    expect(post4([2, 4, 1, 4, 3, 2])).toEqual([3, 2]);
    expect(post4([4, 1, 4, 2, 2, 2])).toEqual([2, 2, 2]);
    expect(post4([3, 4, 3, 2])).toEqual([3, 2]);
  });

  test('notAlone', () => {
    expect(notAlone([1, 2, 3], 2)).toEqual([1, 3, 3]);
    expect(notAlone([1, 2, 3, 2, 5, 2], 2)).toEqual([1, 3, 3, 5, 5, 2]);
    expect(notAlone([3, 4], 3)).toEqual([3, 4]);
    expect(notAlone([3, 3], 3)).toEqual([3, 3]);
    expect(notAlone([1, 3, 1, 2], 1)).toEqual([1, 3, 3, 2]);
    expect(notAlone([3], 3)).toEqual([3]);
    expect(notAlone([], 3)).toEqual([]);
    expect(notAlone([7, 1, 6], 1)).toEqual([7, 7, 6]);
    expect(notAlone([1, 1, 1], 1)).toEqual([1, 1, 1]);
    expect(notAlone([1, 1, 1, 2], 1)).toEqual([1, 1, 1, 2]);
  });

  test('groupNoAdj', () => {
    expect(groupNoAdj(0, [2, 5, 10, 4], 12)).toBe(true);
    expect(groupNoAdj(0, [2, 5, 10, 4], 14)).toBe(false);
    expect(groupNoAdj(0, [2, 5, 10, 4], 7)).toBe(false);
    expect(groupNoAdj(0, [2, 5, 10, 4, 2], 7)).toBe(true);
    expect(groupNoAdj(0, [2, 5, 10, 4], 9)).toBe(true);
    expect(groupNoAdj(0, [10, 2, 2, 3, 3], 15)).toBe(true);
    expect(groupNoAdj(0, [10, 2, 2, 3, 3], 7)).toBe(false);
    expect(groupNoAdj(0, [], 0)).toBe(true);
    expect(groupNoAdj(0, [1], 1)).toBe(true);
    expect(groupNoAdj(0, [9], 1)).toBe(false);
    expect(groupNoAdj(0, [9], 0)).toBe(true);
    expect(groupNoAdj(0, [5, 10, 4, 1], 11)).toBe(true);
  });

  test('groupSum5', () => {
    expect(groupSum5(0, [2, 5, 10, 4], 19)).toBe(true);
    expect(groupSum5(0, [2, 5, 10, 4], 17)).toBe(true);
    expect(groupSum5(0, [2, 5, 10, 4], 12)).toBe(false);
    expect(groupSum5(0, [2, 5, 4, 10], 12)).toBe(false);
    expect(groupSum5(0, [3, 5, 1], 4)).toBe(false);
    expect(groupSum5(0, [3, 5, 1], 5)).toBe(true);
    expect(groupSum5(0, [1, 3, 5], 5)).toBe(true);
    expect(groupSum5(0, [3, 5, 1], 9)).toBe(false);
    expect(groupSum5(0, [2, 5, 10, 4], 7)).toBe(false);
    expect(groupSum5(0, [2, 5, 10, 4], 15)).toBe(true);
    expect(groupSum5(0, [2, 5, 10, 4], 11)).toBe(false);
    expect(groupSum5(0, [1], 1)).toBe(true);
    expect(groupSum5(0, [9], 1)).toBe(false);
    expect(groupSum5(0, [9], 0)).toBe(true);
    expect(groupSum5(0, [], 0)).toBe(true);
  });

  test('groupSumClump', () => {
    expect(groupSumClump(0, [2, 4, 8], 10)).toBe(true);
    expect(groupSumClump(0, [1, 2, 4, 8, 1], 14)).toBe(true);
    expect(groupSumClump(0, [2, 4, 4, 8], 14)).toBe(false);
    expect(groupSumClump(0, [8, 2, 2, 1], 9)).toBe(true);
    expect(groupSumClump(0, [8, 2, 2, 1], 11)).toBe(false);
    expect(groupSumClump(0, [1], 1)).toBe(true);
    expect(groupSumClump(0, [9], 1)).toBe(false);
  });

  test('splitArray', () => {
    expect(splitArray([2, 2])).toBe(true);
    expect(splitArray([2, 3])).toBe(false);
    expect(splitArray([5, 2, 3])).toBe(true);
    expect(splitArray([5, 2, 2])).toBe(false);
    expect(splitArray([1, 1, 1, 1, 1, 1])).toBe(true);
    expect(splitArray([1, 1, 1, 1, 1])).toBe(false);
    expect(splitArray([])).toBe(true);
    expect(splitArray([1])).toBe(false);
    expect(splitArray([3, 5])).toBe(false);
    expect(splitArray([5, 3, 2])).toBe(true);
    expect(splitArray([2, 2, 10, 10, 1, 1])).toBe(true);
    expect(splitArray([1, 2, 2, 10, 10, 1, 1])).toBe(false);
    expect(splitArray([1, 2, 3, 10, 10, 1, 1])).toBe(true);
  });
});
